### PR TITLE
Enumerable Methods

### PR DESCRIPTION
**Pull request by anapdh**
This project consists of an implementation of an own version of following Ruby’s enumerable methods:

- Each;
- Each with index;
- Select;
- All;
- Any;
- None;
- Count;
- Map;
- Inject.

### CODE REVIEW

**Comment by mimipeshy** [comment]
## STATUS : CHANGES REQUIRED :recycle: 

Hey team @anapdh and @elberthcorniell :wave: , you guys did well in this project, I am impressed by your methods. However, before we mark the project as complete let's add a few more functionalities:

- [x]  take a look at the comments I've left directly in your code point_down

That's it, team! raised_hands
Please :pray:  do the changes :technologist:  :technologist: and submit back for a code review when you are done memo
Happy coding! :rocket:

Any questions? ask me on slack @ peris ndanu :woman_technologist:

File: enumerables.rb
Comment (mimipeshy) [inline enumerables.rb:7]: - [ ] enumerables #my_each returns an Enumerator if no block is given

- [ ]  enumerable  #my_each when a block is given when self is an array returns the array itself. On line 4 convert yield to an array, then return self.

File: enumerables.rb
Comment (mimipeshy) [inline enumerables.rb:24]: - [x] enumerables #my_each_with_index returns an enumerator if no block is given

- [x] enumerable #my_each_with_index when a block is given when self is an array returns the array. Again always convert the yield result to an array then return self.
Hint: try and use `size.times do || ` instead of `for i in 0....self.length do`

File: enumerables.rb
Comment (mimipeshy) [inline enumerables.rb:35]: - [x] #my_select returns an enumerator if no block is given

File: enumerables.rb
Comment (mimipeshy) [inline enumerables.rb:55]: - [x] should have working `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))

- [x] my_all should return false if the block returns false or nil::range, also you are not handling classes, regex and patterns correctly.

File: enumerables.rb
Comment (mimipeshy) [inline enumerables.rb:98]: - [x] Same problem not handling classes, regex and patterns is present on my_none? also it should return false if the block ever returns true for all elements::range.

File: enumerables.rb
Comment (mimipeshy) [inline enumerables.rb:122]: - [x] my_count should return the number of items in enum through enumeration::range, and counts the number of items in enum that are equal to item if an argument is given

File: enumerables.rb
Comment (mimipeshy) [inline enumerables.rb:135]: - [x]  my_map should return a new array with the results of running a given block once for every element in enum::range

- [x] enumerables #my_map returns an Enumerator if no block is given

File: enumerables.rb
Comment (mimipeshy) [inline enumerables.rb:159]: - [x] my_inject is not handling blocks and symbols with enum and it should raise a "LocalJumpError" when no block or argument is given.

**Review by mimipeshy — COMMENTED** [review]
For reference I recommend you to read this: https://ruby-doc.org/core-2.6.5/Enumerable.html evey test case is explained over there.

File: enumerables.rb
Comment (mimipeshy) [inline enumerables.rb:11]: - [x]  my_each is not handling ranges and hashes correctly when a block is given.

**Review by mimipeshy — APPROVED** [review]

**Comment by mimipeshy** [comment]
### STATUS : CHANGES REQUIRED :recycle: 
Hey team :wave:  :wave: , its me again :smile: , am impressed that we are making progress, we have corrected some mistakes but we still need to make more changes before approval. I have attached screenshots below :arrow_down: 

- [x] enumerables #my_each when a block is given when self is an array returns the array itself.
It expects an array but your method returns a range.

s11


- [x]  enumerables #my_each when a block is given when self is a hash behaves like #each calls the given block once for each element in self
- It expects a hash, your method returns nil
w
ndex

- [x]  enumerables #my_each_with_index when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- It expects an array but your method returns a range.

w2

- [x]  enumerables #my_each_with_index when a block is given when self is a range behaves like #each_with_index calls the given block once for each element in self
- Your method returns the correct key in the hash but gives the wrong index

w3

- [x] enumerables #my_all? when no block or argument is given returns false when one of the collection members are false or nil
- It expects a false, your method returns a true

- [x]  enumerables #my_any? when no block or argument is given returns false if at least one of the collection is not true

- It expects a false, your method returns a true

- [x] enumerables #my_none? when no block or argument is given returns false only if one of the collection members is true
- It expects a false, your method returns a true

- [x] enumerables #my_count returns the number of items in enum through enumeration::range
- Implement range when counting items

-[ ] enumerables #my_map returns a new array with the results of running a given block once for every element in enum::range
- Your method returns some wrong values

w4

That's it, team! :raised_hands:
Please :pray: do the changes :technologist: :technologist: and submit back for a code review when you are done memo
Happy coding! :rocket:

Any questions? ask me on slack @ peris ndanu :woman_technologist:

File: enumerables.rb
Comment (ebeagusamuel) [inline enumerables.rb]: - [x] similar to what I said above, your method should return self regardless of what class self is. That is how the original method behaves and yours should behave just like that.

File: enumerables.rb
Comment (ebeagusamuel) [inline enumerables.rb]: - [x] this conditional return statement will cause your method not to work as the original enumerable each method. It should return self regardless of what type of class self is.

**Review by ebeagusamuel — CHANGES_REQUESTED** [review]
# Changes Requested :recycle: 
Hello team :wave: 
I can see that you are making progress and more test cases are now passing.
However, there are still some cases that are still failing and you will need to work on them.
I will highlight those cases below.
#### my_each
- [x] When a block is given and self is a range, it should return self(i.e the range) after running the block for each element of the range.
- [x] my_each should be able to work when self is a hash. When a block is given and self is a hash, my_each should be able to run the block for each element in the hash and return the hash.
#### my_each_with_index
- [x] When a block is given and self is a range, it should return self(i.e the range) after running the block for each element of the range.
- [x] when my_each_with_index is passed a block and self is a range, it should be able to run the block for each element of the range.
- [x] when my_each_with_index is passed a block and self is a hash, it should be able to run the block for each element of the hash.
#### my_any?
- [x] when no block or argument is given, my_any? should return false if at least one of the collection of self is not true. i.e if any of the elements of self does not have a true value, then return false.
#### my_map
- [x] When self is a range, my_map should return a new array with the results of running a given block once for every element in the range.
- [x] In a case when both a block and a proc is provided, my_map should run only the proc.
#### my_inject
- [x] when a symbol is specified without an initial value, my_inject should combine each element of the collection by applying the symbol as a named method. i.e `[1, 2, 3, 4].my_inject(:+)` should return 10.
- [x] It should also be able to work on a range i.e `(1..4).my_inject(:+)` should return 10.
- [x] when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method. i.e `[1, 2, 3, 4].my_inject(20, :*)` should return 480.
- [x] it should also work on a range i.e `(1..4).my_inject(20, :*)` should return 480.

In addition, just like the previous TSE suggested, it will be of great help to you if you check the ruby docs and study the enumerable methods to see how they really work. That way you will b be better informed on how to create your own custom methods that will behave like the original methods.
When you are done making the required changes, please request another review.

slack(@ebeagusamuel)

File: enumerables.rb
Comment (elberthcorniell) [reply enumerables.rb]: Hello Samuel, we didn't know it should return self in any case. We were just fixing the test cases that the last TSE showed us were failling. It will be better if we have this testing file so we don't have to ask for a code review to see if everything is good.

**Review by Georjane — CHANGES_REQUESTED** [review]
## Status: Changes Required 🔁 🚫 
### Hi Team @anapdh and @elberthcorniell  🙌🏽 😊

This is really great work.👏👏 You have done so well and made a lot of progress implementing the changes required by the previous TSE.👍 98% of test cases are passing right now👌, we just need to fix two test cases for your code to be working well and meet the minimum requirement for this project to be approved:✔️

### #my_each_with_index
-  Normally, enumerable **#my_each_with_index** when a block is given when self is a range behaves like **#each_with_index**  that calls the given block once for each element in self. The screenshot below shows how each_with_index behaves when self is a range and then how your my_each_with_index behaves when self is a range:

error_enum

- We have a similar issue when the my_each_with_index method is called on a hash. See screenshot below:

error_enum1


Work on these changes and submit for another code review.👍👍
Feel free to reach out to me on slack, **_Witah Georjane_** 🤝

Good luck and Happy Coding 💻

**Review by Srjnnnn — APPROVED** [review]
# Status: Approved :partying_face: :tada: 
Hello team! Congratulations, your project has passed all test cases and approved :muscle: 
Thanks for making the necessary changes.

You can merge this branch and move on to the next project. :100: 
Eyüp Sercan UYGUR

**Pull request merged**
